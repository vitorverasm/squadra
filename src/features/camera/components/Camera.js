import Moment from 'moment';
import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { RNCamera } from 'react-native-camera';
import CustomModal from '../../../components/CustomModal';
import { formatFilename, requestPermission, savePhoto } from '../../../utils/helpers';
import ActionBar from './ActionBar';
import NotAuthMessage from './NotAuthMessage';

class Camera extends Component {
  constructor(props) {
    super(props);
    this.state = {
      writeExternalPermission: false,
      showModal: false
    };
  }

  takePicture = async () => {
    const { writeExternalPermission } = this.state;
    if (this.camera && writeExternalPermission) {
      const { currentTag } = this.props;
      const options = { quality: 0.5, doNotSave: true, base64: true };
      const data = await this.camera.takePictureAsync(options);
      const filename = formatFilename(Moment().format('DD-MM-YY'));
      savePhoto(data.base64, currentTag, filename);
    }
  };

  onCameraReady = () => {
    const permission = requestPermission('storage');
    this.setState({ writeExternalPermission: permission, cameraReady: true });
  };

  closeModal = () => {
    this.setState({ showModal: false });
  };

  renderModal = (content, height) => {
    const { showModal } = this.state;
    return (
      <CustomModal close={this.closeModal} visible={showModal} content={content} height={height} />
    );
  };

  render() {
    const { cameraReady } = this.state;
    return (
      <View style={styles.container}>
        <RNCamera
          ref={(camera) => {
            this.camera = camera;
          }}
          style={styles.preview}
          type={RNCamera.Constants.Type.back}
          flashMode={RNCamera.Constants.FlashMode.off}
          captureAudio={false}
          onStatusChange={({ cameraStatus }) => {
            if (cameraStatus === 'NOT_AUTHORIZED') {
              requestPermission('camera');
            }
          }}
          onCameraReady={this.onCameraReady}
          notAuthorizedView={<NotAuthMessage />}
        />
        {cameraReady ? <ActionBar takePhoto={this.takePicture} /> : null}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    backgroundColor: 'transparent'
  },
  preview: {
    flex: 1
  }
});

export default Camera;
