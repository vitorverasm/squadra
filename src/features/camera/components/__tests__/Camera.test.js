import { shallow, mount } from 'enzyme';
import React from 'react';
import Camera from '../Camera';

const { JSDOM } = require('jsdom');

const jsdom = new JSDOM();
const { window } = jsdom;

function copyProps(src, target) {
  const props = Object.getOwnPropertyNames(src)
    .filter(prop => typeof target[prop] === 'undefined')
    .map(prop => Object.getOwnPropertyDescriptor(src, prop));
  Object.defineProperties(target, props);
}

global.window = window;
global.document = window.document;
global.navigator = {
  userAgent: 'node.js',
};
copyProps(window, global);

// Ignore React Web errors when using React Native
// but still show relevant errors
const suppressedErrors = /(React does not recognize the.*prop on a DOM element|Unknown event handler property|is using uppercase HTML|Received `true` for a non-boolean attribute `accessible`|The tag.*is unrecognized in this browser)|is using incorrect casing|Received `true` for a non-boolean attribute `enabled`/;
const realConsoleError = console.error; // eslint-disable-line
// eslint-disable-next-line
console.error = message => {
  if (message.match(suppressedErrors)) {
    return;
  }
  realConsoleError(message);
};

describe('Camera', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Camera />);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('onCameraReady', () => {
    const wrapper = shallow(<Camera />);
    wrapper.instance().onCameraReady();
    console.log('@wrapper: ', wrapper);
    // expect(wrapper.state().writeExternalPermission).toBe(true);
  });
});
