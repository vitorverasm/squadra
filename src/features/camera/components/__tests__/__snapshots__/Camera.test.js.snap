// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera renders correctly 1`] = `
"<View style={{...}}>
  <RNCamera style={{...}} type=\\"back\\" flashMode={[undefined]} captureAudio={false} onCameraReady={[Function]} pendingAuthorizationView={{...}} />
  <View style={{...}}>
    <TouchableOpacity onPress={[Function: _callee]} style={{...}} activeOpacity={0.2}>
      <Text style={{...}}>
         SNAP 
      </Text>
    </TouchableOpacity>
  </View>
  <CustomModal close={[Function]} visible={false} />
</View>"
`;
